crates/red_knot_python_semantic/resources/mdtest/intersection_types.md

```diff
@@ -635,6 +635,68 @@ def _(
     reveal_type(i8)  # revealed: Never
 ```
 
+### Simplifications of `bool`, `AlwaysTruthy` and `AlwaysFalsy`
+
+In general, intersections with `AlwaysTruthy` and `AlwaysFalsy` cannot be simplified. Naively, you
+might think that `int & AlwaysFalsy` could simplify to `Literal[0]`, but this is not the case: for
+example, the `False` constant inhabits the type `int & AlwaysFalsy` (due to the fact that
+`False.__class__` is `bool` at runtime, and `bool` subclasses `int`), but `False` does not inhabit
+the type `Literal[0]`.
+
+Nonetheless, intersections of `AlwaysFalsy` or `AlwaysTruthy` with `bool` _can_ be simplified, due
+to the fact that `bool` is a `@final` class at runtime that cannot be subclassed.
+
+```py
+from knot_extensions import Intersection, Not, AlwaysTruthy, AlwaysFalsy
+
+def f(
+    a: Intersection[bool, AlwaysTruthy],
+    b: Intersection[bool, AlwaysFalsy],
+    c: Intersection[bool, Not[AlwaysTruthy]],
+    d: Intersection[bool, Not[AlwaysFalsy]],
+    e: Intersection[bool, AlwaysTruthy, str],
+    f: Intersection[bool, AlwaysFalsy, str],
+    g: Intersection[bool, Not[AlwaysTruthy], str],
+    h: Intersection[bool, Not[AlwaysFalsy], str],
+):
+    reveal_type(a)  # revealed: Literal[True]
+    reveal_type(b)  # revealed: Literal[False]
+    reveal_type(c)  # revealed: Literal[False]
+    reveal_type(d)  # revealed: Literal[True]
+
+    # `bool & AlwaysTruthy & str` -> `Literal[True] & str` -> `Never`
+    reveal_type(e)  # revealed: Never
+    reveal_type(f)  # revealed: Never
+    reveal_type(g)  # revealed: Never
+    reveal_type(h)  # revealed: Never
+```
+
+## Simplification of `LiteralString`, `AlwaysTruthy` and `AlwaysFalsy`
+
+Similarly, intersections between `LiteralString`, `AlwaysTruthy` and `AlwaysFalsy` can be
+simplified, due to the fact that a `LiteralString` inhabitant is known to have `__class__` set to
+exactly `str` (and not a subclass of `str`):
+
+```py
+from knot_extensions import Intersection, Not, AlwaysTruthy, AlwaysFalsy
+from typing_extensions import LiteralString
+
+def f(
+    a: Intersection[LiteralString, AlwaysTruthy],
+    b: Intersection[LiteralString, AlwaysFalsy],
+    c: Intersection[LiteralString, Not[AlwaysTruthy]],
+    d: Intersection[LiteralString, Not[AlwaysFalsy]],
+    e: Intersection[AlwaysFalsy, LiteralString],
+    f: Intersection[Not[AlwaysTruthy], LiteralString],
+):
+    reveal_type(a)  # revealed: LiteralString & ~Literal[""]
```

Comment:
Whether this is a "simplification" of `LiteralString & AlwaysTruthy` is of course questionable, since a positive intersection is in some sense "simpler" than a negated intersection. I don't really mind much which one we go with, but I do think it's useful to pick one of them as our "standard representation" of the type, since the two are equivalent.

Neither is _particularly_ user-friendly in its display right now, but users will probably be more familiar with `Literal` types than they are with `AlwaysTruthy` or `AlwaysFalsy`, so I weakly prefer simplifying to `LiteralString & ~Literal[""]` rather than `LiteralString & AlwaysTruthy`

Reply:
> Whether this is a "simplification" of `LiteralString & AlwaysTruthy` is of course questionable

I think it is!

It's potentially not just a question of displaying types to users, I think there's potentially also more we can do with `~Literal[""]`, for example in narrowing.